sklearn/datasets/openml.py

```diff
@@ -0,0 +1,516 @@
+    (e.g., when including [1, 2, 3], the columns will be relabelled to
+    [0, 1, 2])
+
+    Parameters
+    ----------
+    arff_data : tuple
+        A tuple of three lists of equal size; first list indicating the value,
+        second the x coordinate and the third the y coordinate.
+
+    include_columns : list
+        A list of columns to include.
+
+    Returns
+    -------
+    arff_data_new : tuple
+        Subset of arff data with only the include columns indicated by the
+        include_columns argument.
+    """
+    arff_data_new = (list(), list(), list())
+    reindexed_columns = {column_idx: array_idx for array_idx, column_idx
+                         in enumerate(include_columns)}
+    for val, row_idx, col_idx in zip(arff_data[0], arff_data[1], arff_data[2]):
+        if col_idx in include_columns:
+            arff_data_new[0].append(val)
+            arff_data_new[1].append(row_idx)
+            arff_data_new[2].append(reindexed_columns[col_idx])
+    return arff_data_new
+
+
+def _sparse_data_to_array(arff_data, dtype, include_columns):
+    # turns the sparse data back into an array (can't use toarray() function,
+    # as this does only work on numeric data)
+    num_obs = max(arff_data[1]) + 1
+    y_shape = (num_obs, len(include_columns))
+    reindexed_columns = {column_idx: array_idx for array_idx, column_idx
+                         in enumerate(include_columns)}
+    y = np.empty(y_shape, dtype=dtype)
+    for val, row_idx, col_idx in zip(arff_data[0], arff_data[1], arff_data[2]):
+        if col_idx in include_columns:
+            y[row_idx, reindexed_columns[col_idx]] = val
+    return y
+
+
+def _convert_arff_data(arff_data, dtype_x, col_slice_x, dtype_y, col_slice_y):
+    """
+    converts the arff object into the appropriate matrix type (np.array or
+    scipy.sparse.csr_matrix) based on the 'data part' (i.e., in the
+    liac-arff dict, the object from the 'data' key)
+
+    Parameters
+    ----------
+    arff_data : list or dict
+        as obtained from liac-arff object
+
+    dtype_x : type
+        The data type in which the X data should be returned. Preferred:
+        np.float64 or object
+
+    col_slice_x : list
+        The column indices that are sliced from the original array to return
+        as X data
+
+    type_y : type
+        The data type in which the y data should be returned. Preferred:
+        np.float64 or object
+
+    col_slice_y : list
+        The column indices that are sliced from the original array to return
+        as y data
+
+    Returns
+    -------
+    X : np.array or scipy.sparse.csr_matrix
+    y : np.array
+    """
+    if isinstance(arff_data, list):
+        data = np.array(arff_data, dtype=object)
+        X = np.array(data[:, col_slice_x], dtype=dtype_x)
+        y = np.array(data[:, col_slice_y], dtype=dtype_y)
+        return X, y
+    elif isinstance(arff_data, tuple):
+        if dtype_x is not np.float64:
```

Comment:
other numeric types aren't possible?

Reply:
[Arff officially distinguishes](https://www.cs.waikato.ac.nz/ml/weka/arff.html) between numeric, nominal, string and date. It is true that arff can contain integers in the numeric field, but how these are supposed to be parsed is not defined, and up to the arff reader. In order to prevent unanticipated behavior, I enforced the float datatype, but if you prefer we can also add int (at the expense of some code-complexity). In any case, this comment is relevant at current line 478 and on, which invokes this function. I added a line in the comment to clarify this
```
    if all([feature['data_type'] == "numeric" for feature in features_list
            if feature['name'] in data_columns]):
        dtype_x = np.float64
    else:
        dtype_x = object
```

Reply:
FTR, I added:
```
            # note that dtype_x is in {np.float64, object}. In case the data
            # consists of integers, dtype will be np.float64 cf. the arff 
            # specifications
```